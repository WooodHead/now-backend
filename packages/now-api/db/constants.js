/* eslint-disable import/prefer-default-export  */

export const TABLES = {
  ACTIVITY: 'now_template',
  EVENT: 'now_event',
  RSVP: 'now_rsvp',
  MESSAGE: 'now_messages',
  USER: 'now_user',
};
