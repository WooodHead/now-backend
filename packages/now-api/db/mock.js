/* eslint-disable import/no-extraneous-dependencies  */
import ApolloClient from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import schema from '../schema';

export const mockPromise = (resolver, rejecter) =>
  rejecter ? Promise.reject(rejecter) : Promise.resolve(resolver);

export const mocks = {
  scan: () => {
    throw new Error('Unimplemented mock for scan');
  },
  get: () => {
    throw new Error('Unimplemented mock for get');
  },
  put: () => {
    throw new Error('Unimplemented mock for put');
  },
  update: () => {
    throw new Error('Unimplemented mock for update');
  },
  query: () => {
    throw new Error('Unimplemented mock for query');
  },
  getTemplate: () => {
    throw new Error('Unimplemented mock for getTemplate');
  },
};

jest.mock('../db', () => ({
  scan: table => mocks.scan(table),
  get: (table, key) => mocks.get(table, key),
  getTemplate: id => mocks.get('now_template', { id }),
  put: (table, item) => mocks.put(table, item),
  update: (table, key, expr, values, names = undefined) =>
    mocks.update(table, key, expr, values, names),
  query: params => mocks.query(params),
}));

const apolloCache = new InMemoryCache();

export const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({ schema }),
});
