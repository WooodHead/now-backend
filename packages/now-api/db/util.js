/* eslint-disable import/prefer-default-export */
export const parseDate = date =>
  date === null ? null : new Date(parseInt(date, 10));
