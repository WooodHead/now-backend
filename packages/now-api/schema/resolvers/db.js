const AWS = require('aws-sdk');

const dynamoDb = new AWS.DynamoDB.DocumentClient();

const promisify = foo =>
  new Promise((resolve, reject) => {
    foo((error, result) => {
      if (error) {
        reject(error);
      } else {
        resolve(result);
      }
    });
  });

const get = (table, key = {}) =>
  promisify(callback =>
    dynamoDb.get(
      {
        TableName: table,
        Key: key,
      },
      callback
    )
  ).then(response => response.Item);
const scan = table =>
  promisify(callback =>
    dynamoDb.scan(
      {
        TableName: table,
      },
      callback
    )
  ).then(response => response.Items);

const parseDate = date => (date === null ? null : new Date(parseInt(date, 10)));

const transformEvent = event => ({
  id: event.id,
  title: event.title,
  description: event.description,
  limit: event.limit,
  slug: event.slug,
  time: parseDate(event.time),
  chatChannel: event.channel_id,
  duration: event.duration,
  createdAt: parseDate(event.createdAt),
  updatedAt: parseDate(event.updatedAt),
});

const transformEvents = events => events.map(transformEvent);

const transformActivity = activity => ({
  id: activity.id,
  title: activity.title,
  slug: activity.slug,
  description: activity.description,
  limit: activity.limit,
  duration: activity.duration,
  createdAt: parseDate(activity.createdAt),
  updatedAt: parseDate(activity.updatedAt),
});

const transformActivities = events => events.map(transformActivity);

exports.events = () => scan('now').then(transformEvents);

exports.rawEvent = id => get('now', { id });
exports.event = id => exports.rawEvent(id).then(transformEvent);

exports.activities = () => scan('now_table').then(transformActivities);
exports.activity = slug =>
  exports
    .activities()
    .then(activities => activities.find(activity => slug === activity.slug));
