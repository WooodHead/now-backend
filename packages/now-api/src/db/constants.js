/* eslint-disable import/prefer-default-export  */

export const TABLES = {
  ACTIVITY: 'now_template',
  DEVICE: 'now_device',
  EVENT: 'now_event',
  RSVP: 'now_rsvp',
  MESSAGE: 'now_messages',
  USER: 'now_user',
};
