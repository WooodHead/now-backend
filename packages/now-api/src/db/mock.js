/* eslint-disable import/no-extraneous-dependencies  */
import ApolloClient from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';

import schema from '../schema';
import loaders from './loaders';

export const USER_ID = 'aea68a98-591c-11e8-9cdb-872a15ebfd30';
export const mockPromise = (resolver, rejecter) =>
  rejecter ? Promise.reject(rejecter) : Promise.resolve(resolver);

export const mocks = {
  loadMember: jest.fn(),
};

const apolloCache = new InMemoryCache();

let isAdmin = false;

export const setAdmin = b => {
  isAdmin = b;
};

export const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({
    schema,
    context: () => ({
      token: null,
      user: { id: USER_ID },
      loaders: loaders({ currentUserId: USER_ID }),
      scopes: isAdmin ? ['admin'] : [],
    }),
  }),
});

export const newUserClient = (currentUserAuth0Id, currentUserId) =>
  new ApolloClient({
    cache: apolloCache,
    link: new SchemaLink({
      schema,
      context: {
        token: null,
        currentUserAuth0Id,
        loaders: loaders({ currentUserId }),
      },
    }),
  });
