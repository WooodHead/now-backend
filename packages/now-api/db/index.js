import AWS from 'aws-sdk';
import promisify from 'util.promisify';
import { rsvpId } from '../schema/util';
import { TABLES } from './constants';

const dynamoDb = new AWS.DynamoDB.DocumentClient();

const pGet = promisify(dynamoDb.get.bind(dynamoDb));
const pScan = promisify(dynamoDb.scan.bind(dynamoDb));
const pUpdate = promisify(dynamoDb.update.bind(dynamoDb));
const pPut = promisify(dynamoDb.put.bind(dynamoDb));
const pQuery = promisify(dynamoDb.query.bind(dynamoDb));

export const get = (table, key = {}) =>
  pGet({
    TableName: table,
    Key: key,
  }).then(response => response.Item);
export const put = (table, item) =>
  pPut({
    TableName: table,
    Item: item,
  });
export const update = (table, key, expr, values, names = undefined) =>
  pUpdate({
    TableName: table,
    Key: key,
    ExpressionAttributeValues: values,
    ExpressionAttributeNames: names,
    UpdateExpression: expr,
    ReturnValues: 'ALL_NEW',
  });
export const scan = (table, filter = undefined) =>
  pScan({
    TableName: table,
    FilterExpression: filter,
  }).then(response => response.Items);

export const queryRaw = params => pQuery(params);
export const query = params => pQuery(params).then(response => response.Items);

export const getUserRsvpByEvent = (userId, eventId) =>
  get(TABLES.RSVP, { id: rsvpId(eventId, userId) });

export const getActivity = id => get(TABLES.ACTIVITY, { id });

export const getEvent = id =>
  pQuery({
    TableName: TABLES.EVENT,
    KeyConditionExpression: 'id = :id',
    ExpressionAttributeValues: { ':id': id },
    IndexName: 'id-index',
  }).then(response => response.Items[0]);

export { TABLES };
