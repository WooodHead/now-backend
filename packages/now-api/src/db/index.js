import AWS from 'aws-sdk';
import promisify from 'util.promisify';
import { rsvpId } from '../schema/util';
import { TABLES } from './constants';
import { isDev } from '../util';

const dynamoOpts = {};
if (isDev()) {
  dynamoOpts.logger = console;
}

const dynamoDb = new AWS.DynamoDB.DocumentClient(dynamoOpts);

const pGet = promisify(dynamoDb.get.bind(dynamoDb));
const pScan = promisify(dynamoDb.scan.bind(dynamoDb));
const pUpdate = promisify(dynamoDb.update.bind(dynamoDb));
const pPut = promisify(dynamoDb.put.bind(dynamoDb));
const pQuery = promisify(dynamoDb.query.bind(dynamoDb));

export const get = (table, key = {}) =>
  pGet({
    TableName: table,
    Key: key,
  }).then(response => response.Item);

export const put = (table, item, condition = undefined) =>
  pPut({
    TableName: table,
    Item: item,
    ConditionExpression: condition,
  });

export const update = (
  table,
  key,
  expr,
  values,
  names = undefined,
  condition = undefined
) =>
  pUpdate({
    TableName: table,
    Key: key,
    ExpressionAttributeValues: values,
    ExpressionAttributeNames: names,
    UpdateExpression: expr,
    ConditionExpression: condition,
    ReturnValues: 'ALL_NEW',
  });

export const updateDynamic = (table, key, values, condition = undefined) => {
  const expressions = [];
  const expressionNames = {};
  const expressionValues = {};
  Object.entries(values).forEach(([k, v], i) => {
    const attrName = `#f${i}`;
    const fieldName = `:f${i}`;
    expressions.push(`${attrName}=${fieldName}`);
    expressionNames[attrName] = k;
    expressionValues[fieldName] = v;
  });

  return update(
    table,
    key,
    `set ${expressions.join(', ')}`,
    expressionValues,
    expressionNames,
    condition
  );
};

export const scan = (table, filter = undefined) =>
  pScan({
    TableName: table,
    FilterExpression: filter,
  }).then(response => response.Items);

export const queryRaw = params => pQuery(params);
export const query = params => pQuery(params).then(response => response.Items);

export const getUserRsvpByEvent = (userId, eventId) =>
  get(TABLES.RSVP, { id: rsvpId(eventId, userId) });

export const getActivity = id => get(TABLES.ACTIVITY, { id });

export const getEvent = id =>
  pQuery({
    TableName: TABLES.EVENT,
    KeyConditionExpression: 'id = :id',
    ExpressionAttributeValues: { ':id': id },
    IndexName: 'id-index',
  }).then(response => response.Items[0]);

export { TABLES };
