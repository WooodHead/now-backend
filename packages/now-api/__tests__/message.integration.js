import gql from 'graphql-tag';

import { mocks, mockPromise, client } from '../db/mock';

const mockDynamoMessage1 = {
  eventId: '1',
  text: 'message text',
  ts: 1519765415755,
  userId: '1',
};

const mockDynamoMessage2 = {
  eventId: '1',
  text: 'message text',
  ts: 1519765415765,
  userId: '2',
};

const mockDynamoMessage3 = {
  eventId: '1',
  text: 'message text',
  ts: 1519765415775,
  userId: '2',
};

const mockDynamoMessage4 = {
  eventId: '1',
  text: 'message text',
  ts: 1519765415785,
  userId: '2',
};

const mockDynamoMessage5 = {
  eventId: '1',
  text: 'message text',
  ts: 1519765415795,
  userId: '2',
};
mocks.query = () =>
  mockPromise([
    mockDynamoMessage1,
    mockDynamoMessage2,
    mockDynamoMessage3,
    mockDynamoMessage4,
    mockDynamoMessage5,
  ]);
mocks.get = () => mockPromise({ id: 1 });

describe('message', () => {
  it('returns eventMessages', async () => {
    const results = client.query({
      query: gql`
        {
          eventMessages(eventId: "1") {
            edges {
              node {
                event {
                  id
                }
                text
                ts
                user {
                  id
                }
              }
            }
          }
        }
      `,
    });
    const { data } = await results;
    expect(data).toMatchSnapshot();
  });
  it('returns first n results', async () => {
    const results = client.query({
      query: gql`
        {
          eventMessages(eventId: "1", first: 2) {
            edges {
              node {
                event {
                  id
                }
                text
                ts
                user {
                  id
                }
              }
            }
          }
        }
      `,
    });
    const { data } = await results;
    expect(data).toMatchSnapshot();
  });
  it('returns last n results', async () => {
    const results = client.query({
      query: gql`
        {
          eventMessages(eventId: "1", last: 2) {
            edges {
              node {
                event {
                  id
                }
                text
                ts
                user {
                  id
                }
              }
            }
          }
        }
      `,
    });
    const { data } = await results;
    expect(data).toMatchSnapshot();
  });
});
