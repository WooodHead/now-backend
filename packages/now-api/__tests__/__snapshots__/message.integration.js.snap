// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message returns eventMessages 1`] = `
Object {
  "eventMessages": Object {
    "__typename": "EventMessagesConnection",
    "edges": Array [
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415755",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.0.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.0",
      },
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415765",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.1.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.1",
      },
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415775",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.2.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.2",
      },
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415785",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.3.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.3",
      },
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415795",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.4.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"}).edges.4",
      },
    ],
    Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\"})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`message returns first n results 1`] = `
Object {
  "eventMessages": Object {
    "__typename": "EventMessagesConnection",
    "edges": Array [
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415755",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"first\\":2}).edges.0.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"first\\":2}).edges.0",
      },
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415765",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"first\\":2}).edges.1.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"first\\":2}).edges.1",
      },
    ],
    Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"first\\":2})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`message returns last n results 1`] = `
Object {
  "eventMessages": Object {
    "__typename": "EventMessagesConnection",
    "edges": Array [
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415785",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"last\\":2}).edges.0.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"last\\":2}).edges.0",
      },
      Object {
        "__typename": "EventMessagesEdge",
        "node": Object {
          "__typename": "Message",
          "event": Object {
            "__typename": "Event",
            "id": "1",
            Symbol(id): "Event:1",
          },
          "text": "message text",
          "ts": "1519765415795",
          "user": null,
          Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"last\\":2}).edges.1.node",
        },
        Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"last\\":2}).edges.1",
      },
    ],
    Symbol(id): "$ROOT_QUERY.eventMessages({\\"eventId\\":\\"1\\",\\"last\\":2})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;
